examples/spin-timer/src/lib.rs

```diff
@@ -113,15 +118,30 @@ impl TriggerExecutor for TimerTrigger {
         }
         Ok(())
     }
+
```

Comment:
Per my comment on the issue, let's find a way that doesn't require every custom trigger to copy this code.  (That might be as simple as defaulting things in the trait, I don't know.)

Reply:
Sadly this can't be defaulted as-is because you can't have associated type defaults on traits (https://github.com/rust-lang/rust/issues/29661).

Reply:
@alexcrichton Would it be reasonable to add traits e.g. `spin_core::SpinInstance(Pre)` and have these methods return `impl SpinInstance(Pre)`?

Reply:
I suppose the `SpinInstance` trait might be a bit gnarly there and could end up boiling down to a component-or-module enum in disguise. Depends on what exactly you're hoping to enable here.

Reply:
Sounds like a reasonable concern to me! I ended up creating a `TriggerInstancePre` trait for the `spin-trigger` trait before I came here and read @lann's comments. I think that this should have the desired ergonomics but I'm happy to move/rename the trait too

Reply:
I think given the amount of cargo-culty copying required to make a custom trigger today, this is fine. The whole interaction between spin-core, spin-app, and component instantiation is pretty wonky and became wonkier with the (partial) migration to components, and I don't think this trigger architecture can or should survive long-term.

Probably more important in the near-term would be to actually document (mea culpa) this interface a bit to help people along.

Reply:
If I read this correctly, trigger authors would no longer cargo-cult the `let EitherInstance::Component...` fragment and instead cargo-cult the `type InstancePre = ...` fragment.  That feels like less cargo-culting and I'm in favour.  Thank you Alex for simplifying this!

(If I've misread it, though, @alexcrichton, please hold off so I can get my head around the actual experience.)

Reply:
No yep you've read correctly, and once Rust supports defaults for associated types we can remove the `type InstancePre = ...` as well (or at least that's my understanding of associated typ defaults)